tensorflow/compiler/mlir/quantization/stablehlo/passes/bridge/convert_mhlo_quant_to_int.cc

```diff
@@ -1215,20 +1222,10 @@ FailureOr<DotLikeDimensionNumbers> VerifyAndConstructDims(
   auto dims = op.getDimensionNumbers();
   if (IsConvNHWC(dims)) {
     // 2D Convolution.
-    return DotLikeDimensionNumbers{/*lhs_batching_dims=*/{0},
-                                   /*lhs_spatial_dims=*/{1, 2},
-                                   /*lhs_contracting_dims=*/{3},
-                                   /*rhs_batching_dims=*/{},
-                                   /*rhs_spatial_dims=*/{0, 1},
-                                   /*rhs_contracting_dims=*/{2}};
+    return true;
   } else if (IsConvNDHWC(dims)) {
     // 3D Convolution.
-    return DotLikeDimensionNumbers{/*lhs_batching_dims=*/{0},
```

Comment:
IIUC, the issue is that these array values are not being stored in constant memory, making it invalid to create an array ref to them?

This is a strange case since usually all data is owned by an MLIRContext so we can create array refs to anything. In this case we're dealing with a mix of some things owned by MLIRContext and some on the stack. With that said, it may be better to have this struct own its own data. Could we change `DotLikeDimensionNumbers` to have copies of the numbers, they're all very small or empty, I don't think it would cause much of a performance hit:

```
struct DotLikeDimensionNumbers {
  SmallVector<int64_t> lhs_batching_dims;
  SmallVector<int64_t> lhs_spatial_dims;
  SmallVector<int64_t> lhs_contracting_dims;
  SmallVector<int64_t> rhs_batching_dims;
  SmallVector<int64_t> rhs_spatial_dims;
  SmallVector<int64_t> rhs_contracting_dims;
};
```

Then we shouldn't need to change anything else (except for using `to_vector` at one `DotLike` ctor callsite), as the memory is owned by DotLikeDimensionsNumbers and should be valid for the llvm concat calls.

Reply:
I have updated the PR as you suggested and while it no longer triggers the address sanitizer I am still worried about the use of llvm::concat as it seemed to be creating a range rather than joining the contents of two arrays.

Reply:
@GleasonK the concat is indeed creating a range from the two arguments being treated as int64_t but considering that they are in fact SmallVector arguments this seems very strange to me.
```
794	    SmallVector<int64_t> reduction_dims = to_vector(llvm::concat<const int64_t>(
(gdb) s
llvm::concat<long const, llvm::SmallVector<long, 6u> const&, llvm::SmallVector<long, 6u> const&> ()
    at external/llvm-project/llvm/include/llvm/ADT/STLExtras.h:1189
1189	      std::forward<RangeTs>(Ranges)...);
(gdb) 
llvm::detail::concat_range<long const, llvm::SmallVector<long, 6u> const&, llvm::SmallVector<long, 6u> const&>::concat_range (
    Ranges#1=..., Ranges#0=..., this=0xffffdc75c560) at external/llvm-project/llvm/include/llvm/ADT/STLExtras.h:1189
1189	      std::forward<RangeTs>(Ranges)...);
(gdb) 
std::tuple<llvm::SmallVector<long, 6u> const&, llvm::SmallVector<long, 6u> const&>::tuple<true, true> (__a2=..., __a1=..., 
    this=0xffffdc75c560) at external/llvm-project/llvm/include/llvm/ADT/STLExtras.h:1189
1189	      std::forward<RangeTs>(Ranges)...);
(gdb) 
std::_Tuple_impl<0ul, llvm::SmallVector<long, 6u> const&, llvm::SmallVector<long, 6u> const&>::_Tuple_impl (__tail#0=..., __head=..., 
    this=0xffffdc75c560) at external/llvm-project/llvm/include/llvm/ADT/STLExtras.h:1189
1189	      std::forward<RangeTs>(Ranges)...);
(gdb) 
std::_Tuple_impl<1ul, llvm::SmallVector<long, 6u> const&>::_Tuple_impl (__head=..., this=0xffffdc75c560)
    at external/llvm-project/llvm/include/llvm/ADT/STLExtras.h:1189
1189	      std::forward<RangeTs>(Ranges)...);
(gdb) 
std::_Head_base<1ul, llvm::SmallVector<long, 6u> const&, false>::_Head_base (__h=..., this=0xffffdc75c560)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/tuple:969
969		tuple(const _T1& __a1, const _T2& __a2)
(gdb) 
std::_Tuple_impl<0ul, llvm::SmallVector<long, 6u> const&, llvm::SmallVector<long, 6u> const&>::_Tuple_impl (__tail#0=..., __head=..., 
    this=0xffffdc75c560) at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/tuple:217
217	      : _Inherited(__tail...), _Base(__head) { }
(gdb) 
std::_Head_base<0ul, llvm::SmallVector<long, 6u> const&, false>::_Head_base (__h=..., this=0xffffdc75c568)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/tuple:129
129	      constexpr _Head_base(const _Head& __h)
(gdb) 
llvm::to_vector<llvm::detail::concat_range<long const, llvm::SmallVector<long, 6u> const&, llvm::SmallVector<long, 6u> const&> > (
    Range=...) at external/llvm-project/llvm/include/llvm/ADT/SmallVector.h:1308
1308	  return {std::begin(Range), std::end(Range)};
```

Reply:
Is this still producing nonsense when the values are vectors? It looks like `llvm::concat` returns a concat_iterator which should take into account that the ranges are not contiguous, which I think would be safe if we're then converting them to a vector?

If not, then we can always implement a vector concat using vector's `insert` method if needed. Will leave that up to you. The change LGTM though assuming if fixes the issues you were previously encountering

Reply:
Sorry, yes, concat is producing sane results, so all good. Thank you.